typescript.md

```diff
@@ -0,0 +1,283 @@
+# Guidelines for typing in TypeScript
+
+## Where to store types specific to a repository
+
+When types are shared between modules and used sparsely, create a `src/types.ts` file
+as per TypeScript's conventions, where you will pile up all the common types.
+
+Presumably you will also build a declaration at `lib/types.d.ts`.
+If your repository hosts a library, then it will be useful to have those types easily imported into parent modules.
+Consider adding `types.d.ts` in the root of your repository with the content `export * from './lib/types.d.ts';`,
+since this will make parent modules be able to write `import {FunkyType} from 'funky-module/types';`.
+
+
+## Casting and guards
+
+At times, you may need to clarify to typescript that a variable is certainly of a specific type within a block.
+
+```typescript
+let fn = function(a: boolean | string): boolean {
+  if ((a as string).match) {
+    a = a as string;
+    return a.match(/^true$/) === null;
+  }
+  return a;
+};
+```
+
+Similarly, you may use:
+
+* the `in` operator e.g. `if ('match' in a)`
+* the `typeof` guard e.g. `if (typeof a === 'string')`
+* the `instanceof` guard e.g. `let a = new String('foo'); if (a instanceof String`
+* `if ((a as any).match) { a = a as string; ...`
+* `if ((a as any).match) { a = a as unknown as string; ...` (in extreme cases)
+
+[Read more about this.](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types)
+
+
+## Use type assertions to `unknown` when the going gets tough
+
+At times, you may hit issues because of type mismatching.
+
+If you are certain that the implementation is correct (but maybe the typing is only partially correct),
+rather than fall back to `// @ts-ignore`, use a type assertion to unknown and then to the intended type.
+
+```typescript
+type Person = {
+  name: string;
+  gender: string;
+};
+
+let fn = function(): Person {
+  let p = {} as Partial<Person>;
```

Comment:
Room for improvement? Using `Partial` is a good example but might miss the point if the reader is not familiar with `Partial`. Simpler example? Explaining Partial before this section?